pkg/jetstream/consume.go

```diff
@@ -0,0 +1,134 @@
+package jetstream
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/nats-io/nats.go/jetstream"
+)
+
+type ConsumerBuilder func(ctx context.Context, js jetstream.JetStream, topic string) (
+	jetstream.Consumer,
+	func(context.Context, watermill.LoggerAdapter),
+	error)
+
+type ConsumerNamer func(string, string) string
+
+func defaultConsumerNamer(topic string, group string) string {
+	if group != "" {
+		return fmt.Sprintf("watermill__%s__%s", group, topic)
+	}
+	return fmt.Sprintf("watermill__%s", topic)
+}
+
+// ExistingConsumer is used to connect to a stream/consumer that already exist with the given topic name - it will not attempt creation of any broker-managed resources.
+// It takes as an argument a function to transform the topic into a consumer name, passing nil will invoke the default behavior
+// consumerName := fmt.Sprintf("watermill__%s", topic)
+func ExistingConsumer(consumerNamer ConsumerNamer, group string) ConsumerBuilder {
+	return func(ctx context.Context, js jetstream.JetStream, topic string) (jetstream.Consumer, func(context.Context, watermill.LoggerAdapter), error) {
+		stream, err := js.Stream(ctx, topic)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to get stream for topic %s: %w", topic, err)
+		}
+
+		var consumerName string
+		if consumerNamer == nil {
+			consumerName = fmt.Sprintf("watermill__%s", topic)
+		} else {
+			consumerName = consumerNamer(topic, group)
+		}
+
+		consumer, err := stream.Consumer(ctx, consumerName)
+
+		return consumer, nil, err
+	}
+}
+
+// GroupedConsumer builds a callback to create a consumer in the given group.
+// The closing function is not returned since a single subscription in the group cannot know when the backing consumer should be deleted.
+func GroupedConsumer(groupName string) ConsumerBuilder {
+	return func(ctx context.Context, js jetstream.JetStream, topic string) (jetstream.Consumer, func(context.Context, watermill.LoggerAdapter), error) {
+		consumer, _, err := createOrUpdateConsumerWithCloser(ctx, js, topic, groupName, defaultConsumerNamer)
+		return consumer, nil, err
+	}
+}
+
+// EphemeralConsumer builds a callback to create a consumer, returning a function that will be used to delete the broker-managed consumer.
+func EphemeralConsumer() ConsumerBuilder {
+	return func(ctx context.Context, js jetstream.JetStream, topic string) (jetstream.Consumer,
+		func(context.Context, watermill.LoggerAdapter),
+		error) {
+		return createOrUpdateConsumerWithCloser(ctx, js, topic, watermill.NewShortUUID(), defaultConsumerNamer)
+	}
+}
```

Comment:
@roblaszczak we talked about this some last time and this new jetstream API makes it a lot simpler I think.  Not perfect and it probably makes sense to allow a custom namer for any of them.  But it allows a ton of flexibility in how you dynamically provision and cleanup any JS resources without a ton of thought about when it needs to happen.

Will revisit before we merge to minimize chances of changing the API once it's out in the wild.